OJS.Workers.ExecutionStrategies/RunSpaAndExecuteMochaTestsExecutionStrategy.cs

```diff
@@ -0,0 +1,389 @@
+                                test))
+                            .SelectMany(resultList => resultList));
+            return result;
+        }
+
+        private ICollection<TestResult> RunIndividualTest(
+            string codeSavePath,
+            IExecutor executor,
+            IExecutionContext<TestsInputModel> executionContext,
+            TestContext test)
+        {
+            var processExecutionResult = this.Execute(executionContext, executor, codeSavePath, test.Input);
+            return this.ExtractTestResultsFromReceivedOutput(processExecutionResult.ReceivedOutput, test.Id);
+        }
+
+        protected override IExecutor CreateExecutor()
+            => this.CreateExecutor(ProcessExecutorType.Standard);
+
+        private void ExtractSubmissionFiles<TInput>(IExecutionContext<TInput> executionContext)
+        {
+            this.ValidateAllowedFileExtension(executionContext);
+
+            var submissionFilePath = FileHelpers.BuildPath(this.WorkingDirectory, "temp");
+            File.WriteAllBytes(submissionFilePath, executionContext.FileContent);
+            using (ZipFile zip = ZipFile.Read(submissionFilePath))
+            {
+                zip.RemoveSelectedEntries("node_modules/*");
+                zip.Save();
+            }
+
+            FileHelpers.RemoveFilesFromZip(submissionFilePath, RemoveMacFolderPattern);
+            FileHelpers.UnzipFile(submissionFilePath, this.UserApplicationPath);
+
+            Directory.CreateDirectory(this.TestsPath);
+            Directory.CreateDirectory(this.NginxConfFileDirectory);
+        }
+
+        private void ValidateAllowedFileExtension<TInput>(IExecutionContext<TInput> executionContext)
+        {
+            var trimmedAllowedFileExtensions = executionContext.AllowedFileExtensions?.Trim();
+            var allowedFileExtensions = (!trimmedAllowedFileExtensions?.StartsWith(".") ?? false)
+                ? $".{trimmedAllowedFileExtensions}"
+                : trimmedAllowedFileExtensions;
+
+            if (allowedFileExtensions != Constants.ZipFileExtension)
+            {
+                throw new ArgumentException("Submission file is not a zip file!");
+            }
+        }
+
+        private ICollection<TestResult> ExtractTestResultsFromReceivedOutput(string receivedOutput, int parentTestId)
+        {
+            JsonExecutionResult mochaResult = JsonExecutionResult.Parse(this.PreproccessReceivedExecutionOutput(receivedOutput));
+            if (mochaResult.TotalTests == 0)
+            {
+                return new List<TestResult>
+                {
+                    new TestResult
+                    {
+                        Id = parentTestId,
+                        IsTrialTest = false,
+                        ResultType = TestRunResultType.WrongAnswer,
+                        CheckerDetails = new CheckerDetails
+                        {
+                            UserOutputFragment = receivedOutput
+                        }
```

Comment:
Why add any test result if `TotalTests` are 0?

Reply:
This is a workaround for mocha results. Otherwise the result got displayed as a changed tests message or compile time error, i dont remember anymore.